x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/normalizer/ShortCircuitingRenormalizer.java

```diff
@@ -88,61 +83,23 @@ public void shutdown() {
         waitUntilIdle();
     }
 
-    private Quantiles getEarliestQuantiles() {
-        QuantilesWithLatch earliestQuantilesWithLatch = quantilesDeque.peekFirst();
-        return (earliestQuantilesWithLatch != null) ? earliestQuantilesWithLatch.getQuantiles() : null;
-    }
-
-    private QuantilesWithLatch getLatestQuantilesWithLatchAndClear() {
-        // We discard all but the latest quantiles
-        QuantilesWithLatch latestQuantilesWithLatch = null;
-        for (QuantilesWithLatch quantilesWithLatch = quantilesDeque.pollFirst(); quantilesWithLatch != null; quantilesWithLatch =
-            quantilesDeque.pollFirst()) {
-            // Count down the latches associated with any discarded quantiles
-            if (latestQuantilesWithLatch != null) {
-                latestQuantilesWithLatch.getLatch().countDown();
-            }
-            latestQuantilesWithLatch = quantilesWithLatch;
-        }
-        return latestQuantilesWithLatch;
+    private synchronized AugmentedQuantiles getLatestAugmentedQuantilesAndClear() {
+        AugmentedQuantiles latest = latestQuantilesHolder;
+        latestQuantilesHolder = null;
+        return latest;
     }
 
     private boolean tryStartWork() {
         return semaphore.tryAcquire();
     }
 
-    private boolean tryFinishWork() {
-        // We cannot tolerate new work being added in between the isEmpty() check and releasing the semaphore
-        synchronized (quantilesDeque) {
-            if (quantilesDeque.isEmpty() == false) {
-                return false;
-            }
-            semaphore.release();
-            return true;
-        }
-    }
-
-    private void forceFinishWork() {
-        // We cannot allow new quantiles to be added while we are failing from a previous renormalization failure.
-        synchronized (quantilesDeque) {
-            // We discard all but the earliest quantiles, if they exist
-            QuantilesWithLatch earliestQuantileWithLatch = null;
-            for (QuantilesWithLatch quantilesWithLatch = quantilesDeque.pollFirst(); quantilesWithLatch != null; quantilesWithLatch =
-                quantilesDeque.pollFirst()) {
-                if (earliestQuantileWithLatch == null) {
-                    earliestQuantileWithLatch = quantilesWithLatch;
-                }
-                // Count down all the latches as they no longer matter since we failed
-                quantilesWithLatch.latch.countDown();
-            }
-            // Keep the earliest quantile so that the next call to doRenormalizations() will include as much as the failed normalization
-            // window as possible.
-            // Since this latch is already countedDown, there is no reason to put it in the `latchDeque` again
-            if (earliestQuantileWithLatch != null) {
-                quantilesDeque.addLast(earliestQuantileWithLatch);
-            }
-            semaphore.release();
+    private synchronized boolean tryFinishWork() {
```

Comment:
I've preserved the old functionality:

- It doesn't remember the earliest time if a renormalization fails
- It does remember the earliest time if we discard a quantiles document because we got a newer one before the earlier one was processed

Failed renormalizations should be rare so I don't think this will make a difference very often. I'd rather just make the one change in this PR. If there is a user who repeatedly gets failed renormalizations then I don't want to introduce another build-up of work by remembering the earliest time forever.